crates/router/src/connector/shift4/transformers.rs

```diff
@@ -87,49 +125,94 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for Shift4PaymentsRequest {
+                    amount: item.request.amount.to_string(),
+                    card: Some(CardPayment::CardToken(card_token.id)),
+                    currency: item.request.currency.to_string(),
+                    description: item.description.clone(),
+                    captured: item.request.is_auto_capture(),
+                    payment_method: None,
+                    flow: None,
+                })))
+            }
+            _ => Err(errors::ConnectorError::NotImplemented("Payment Method".to_string()).into()),
+        }
+    }
+}
+
+fn get_card_payment_request<T>(
+    item: &types::RouterData<T, types::PaymentsAuthorizeData, types::PaymentsResponseData>,
     card: &api_models::payments::Card,
 ) -> Result<Shift4PaymentsRequest, Error> {
-    let submit_for_settlement = submit_for_settlement(item);
-    let card = Some(Card {
+    let submit_for_settlement = item.request.is_auto_capture();
+    let card = Card {
         number: card.card_number.clone(),
         exp_month: card.card_exp_month.clone(),
         exp_year: card.card_exp_year.clone(),
         cardholder_name: card.card_holder_name.clone(),
-    });
-    Ok(Shift4PaymentsRequest {
-        amount: item.request.amount.to_string(),
-        card,
-        currency: item.request.currency.to_string(),
-        description: item.description.clone(),
-        captured: submit_for_settlement,
-        payment_method: None,
-        flow: None,
-    })
+    };
+    if item.is_three_ds() {
+        Ok(Shift4PaymentsRequest::ThreeDSRequest(Shift43DSRequest {
+            amount: item.request.amount.to_string(),
+            currency: item.request.currency.to_string(),
+            card_number: card.number,
+            card_exp_month: card.exp_month,
+            card_exp_year: card.exp_year,
+            return_url: item
+                .request
+                .complete_authorize_url
+                .clone()
+                .ok_or_else(|| errors::ConnectorError::RequestEncodingFailed)?,
+        }))
+    } else {
+        Ok(Shift4PaymentsRequest::Non3DSRequest(Box::new(
+            Shift4Non3DSRequest {
+                amount: item.request.amount.to_string(),
+                card: Some(CardPayment::RawCard(card)),
+                currency: item.request.currency.to_string(),
+                description: item.description.clone(),
+                captured: submit_for_settlement,
+                payment_method: None,
+                flow: None,
+            },
+        )))
+    }
 }
 
-fn get_bank_redirect_request(
-    item: &types::PaymentsAuthorizeRouterData,
+fn get_bank_redirect_request<T>(
+    item: &types::RouterData<T, types::PaymentsAuthorizeData, types::PaymentsResponseData>,
     redirect_data: &payments::BankRedirectData,
 ) -> Result<Shift4PaymentsRequest, Error> {
-    let submit_for_settlement = submit_for_settlement(item);
+    let submit_for_settlement = item.request.is_auto_capture();
```

Comment:
We have 4 CaptureMethods `Automatic, Manual, ManualMultiple, Scheduled,`. 
Implementation for each capture type may differ based on connector or they may not support some capture type. Here because of `request.is_auto_capture()` method we are considering whether it is `Automatic` or `Manual`, we are ignoring the other 2 capture types (Even though right now we don't have support for these 2, future this will requires lot of refactoring). So instead of this why Can't we just have case condition for `Automatic`, `Manual` rest all others throw FlowNotSupported?

Reply:
When you want to check other type add one more method to it like `is_manual_multiple` so instead of writing a switch case in each connector like how it is done here, we can use the same logic in all the places. i don't understand why you are mentioning `will requires lot of refactoring`

Reply:
When you want to check other type add one more method to it like `is_manual_multiple` so instead of writing a switch case in each connector like how it is done here, we can use the same logic in all the places. i don't understand why you are mentioning `will requires lot of refactoring`

Reply:
First concern here is if we receive `CaptureMethods` other that `Automatic ` or `Manual ` we should propagate error. Instead you are allowing the payment .
Secondly I don't agree with keep on adding method for each CaptureMethods. As to decide the CaptureMethods we have to call all the methods like `is_auto_capture`, `is_manual_multiple`, `is_scheduled`. Which is nothing but the case matching like how we are doing for PaymentMethods)
Yes refactor would be required for the connectors which uses `is_auto_capture` method when we start to support new CaptureMethods. But we can skip that refactor if we handle `CaptureMethods ` in match statement within every connector code.

Reply:
Currently we have 
```
fn is_auto_capture(&self) -> bool {
        matches!(
            self.capture_method,
            Some(storage_models::enums::CaptureMethod::Automatic) | None
        )
    }
```
Tell me how this will change in future if we have one more new `Capture Method`
@ArjunKarthik

Reply:
Even now we have 4 values for CaptureMethod enum,

> pub enum CaptureMethod {
    #[default]
    Automatic,
    Manual,
    ManualMultiple,
    Scheduled,
}

Reply:
Yes , even though it has 4 capture methods, it did not affect any other methods and served the auto_capture check too. so why should there be a change in future is required?
```
fn is_auto_capture(&self) -> bool {
        matches!(
            self.capture_method,
            Some(storage_models::enums::CaptureMethod::Automatic) | None
        )
    }
 ```